docs/zynq-7000-getting-started-guide.md

```diff
@@ -0,0 +1,183 @@
+```
+
+## Setup and Compilation
+
+Run these commands to clone and build the necessary repos (`~/cosim` assumed as the base directory).
+
+### Create the base directory
+
+```bash
+mkdir ~/cosim
+```
+
+### SystemC Setup
+
+```bash
+# Build and install the binaries
+cd ~/cosim
+SYSC_VERSION=systemc-2.3.2
+wget https://www.accellera.org/images/downloads/standards/systemc/systemc-2.3.2.tar.gz
+tar xf systemc.2.3.2.tar.gz
+cd systemc-2.3.2/
+CXXFLAGS=-std=c++11 ./configure ; make
+sudo make install
+```
+
+### QEMU Setup
+
+```bash
+cd ~/cosim
+git clone https://github.com/Xilinx/qemu.git
+cd qemu
+git checkout 7e3e3ae09af8abe70383a5
+git submodule update --init dtc
+
+# Configure and build
+./configure --target-list="arm-softmmu,aarch64-softmmu,microblazeel-softmmu" --enable-fdt --disable-kvm --disable-xen
+make -j$((`nproc`+1))
+sudo make install
+export SYSTEM=/cosim/systemc-2.3.2/
+```
+
+### Demo Setup
+
+```bash
+# Clone the repo and the submodules
+cd ~/cosim
+git clone https://github.com/Xilinx/systemctlm-cosim-demo.git
+cd systemctlm-cosim-demo
+sed -i 's~../libsystemctlm-soc.git~https://github.com/Xilinx/libsystemctlm-soc.git~g' .gitmodules
+git submodule sync --recursive && git submodule update --init libsystemctlm-soc
+
+# Create the Makefile configeration
+cat << EOF | tee .config.mk
+CXXFLAGS=-std=c++11
+SYSTEMC=${HOME/cosim}
+HAVE_VERILOG=n
+HAVE_VERILOG_VERILATOR=n
+HAVE_VERILOG_VCS=n
+VM_TRACE=1
+EOF*
+
+# Build the demo
+make zynq_demo
+```
+
+### Buildroot Setup
+
+```bash
+cd ~/cosim
+git clone https://github.com/buildroot/buildroot.git
+cd buildroot
+git checkout 36edacce9c2c3b90f9bb11
+mkdir handles
+```
+
+For this walkthough we will be configuring the demo to emulate the Zynq-7000 SoC. 
+
+```bash
+# Pull the .dtsi files for the Zynq-7000
+wget https://raw.githubusercontent.com/Xilinx/qemu-devicetrees/master/zynq-pl-remoteport.dtsi
+wget https://raw.githubusercontent.com/Xilinx/linux-xlnx/master/arch/arm/boot/dts/zynq-7000.dtsi
+wget https://raw.githubusercontent.com/Xilinx/device-tree-xlnx/master/device_tree/data/kernel_dtsi/2017.3/BOARD/zc702.dtsi
+
+# Setup the Buildroot compilation configeration
+rm .config .config.old  -f
+cat << EOF | tee ".config"
+BR2_arm=y
+BR2_cortex_a9=y
+BR2_ARM_ENABLE_NEON=y
+BR2_ARM_ENABLE_VFP=y
+BR2_TOOLCHAIN_EXTERNAL=y
+BR2_TOOLCHAIN_EXTERNAL_BOOTLIN=y
+BR2_TOOLCHAIN_EXTERNAL_BOOTLIN_ARMV7_EABIHF_GLIBC_STABLE=y
+BR2_TARGET_GENERIC_GETTY_PORT="ttyPS0"
+BR2_LINUX_KERNEL=y
+BR2_LINUX_KERNEL_CUSTOM_VERSION=y
+BR2_LINUX_KERNEL_CUSTOM_VERSION_VALUE="5.4.75"
+BR2_LINUX_KERNEL_USE_ARCH_DEFAULT_CONFIG=y
+BR2_LINUX_KERNEL_UIMAGE=y
+BR2_LINUX_KERNEL_UIMAGE_LOADADDR="0x8000"
+BR2_LINUX_KERNEL_DTS_SUPPORT=y
+BR2_LINUX_KERNEL_CUSTOM_DTS_PATH="zynq-7000.dtsi  zynq-zc702.dts zynq-pl-remoteport.dtsi"
+BR2_TARGET_ROOTFS_CPIO=y
+BR2_TARGET_ROOTFS_CPIO_GZIP=y
+BR2_TARGET_ROOTFS_EXT2=y
+#BR2_TARGET_ROOTFS_TAR is not set
+EOF
+
+# Build
+make clean && make olddefconfig
```

Comment:
Petalinux pre-builts has the default cpufreq governor configured to 'userspace' in the kernel. For configuring this after above line do:

```
$ make linux-menuconfig
    CPU Power Management  --->
        CPU Frequency scaling  --->
            Default CPUFreq governor (ondemand)  --->

                   ***Select userspace***

```
In case it is done after the brmake command one can rebuild the kernel+dtb only in buildroot with 'make linux-rebuild'.

Without this I got a lot of below prints:

```
[    6.379055] cpu cpu0: _generic_set_opp_clk_only: failed to set clock rate: -16
[    6.379083] cpufreq: __target_index: Failed to change cpu frequency: -16
```

Reply:
If I am understanding this correctly, while a misconfiguration, it does not stop the proper execution of the demo. I am going to look into altering this in an automated fashion when I do my independent verification of the code.

Reply:
Also, since this is the Xilinx fork of QEMU, does it assume the defaults of PetaLinux and thus this arises since we have chosen Buildroot instead?

Reply:
Hi Matt,

Its a no on the question above (Xilinx QEMU is not tailored for the petalinux build), I did not look into the issue above further though so not completly sure why the print comes out at the moment. A solution for the problem (that worked for me) could be to use the [xiinx_zynq_defconfig](https://github.com/Xilinx/linux-xlnx/blob/master/arch/arm/configs/xilinx_zynq_defconfig) when building the kernel as below:

After the dts/dtsi wget fetches also wget the xiinx_zynq_defconfig inside the ~/cosim/buildroot directory.
```
$ wget https://raw.githubusercontent.com/Xilinx/linux-xlnx/3fbb3a6c57a43342a7daec3bbae2f595c50bc969/arch/arm/configs/xilinx_zynq_defconfig
```

Remove BR2_LINUX_KERNEL_USE_ARCH_DEFAULT_CONFIG=y from the 'cat' command creating the buildroot config and add in below instead into the cat command:
```
$ cat << EOF | tee ".config"
...
BR2_LINUX_KERNEL_USE_CUSTOM_CONFIG=y
BR2_LINUX_KERNEL_CUSTOM_CONFIG_FILE="xilinx_zynq_defconfig"
...
EOF
```
(If you already have another solution mind that is ok aswell!)

Best regards,
Francisco